baseapp/router.go

```diff
@@ -1,21 +1,25 @@
 package baseapp
 
 import (
+	"encoding/json"
+	"fmt"
 	"regexp"
 
 	sdk "github.com/cosmos/cosmos-sdk/types"
 )
 
 // Router provides handlers for each transaction type.
 type Router interface {
-	AddRoute(r string, h sdk.Handler) (rtr Router)
+	AddRoute(r string, h sdk.Handler, i sdk.InitGenesis) (rtr Router)
```

Comment:
I'm not sure about this.

What makes InitGenesis different from BeginBlock and EndBlock that it gets this special treatment ? 

See also my thoughts in https://github.com/cosmos/cosmos-sdk/issues/559 (which I don't think Jae agrees with, but anyways).

In the absence of doing this, modules should export methods that can be called by the InitChain function that is stitched together by the application and passed to SetInitChain

Reply:
The problem mostly lies routing the correct genesis option to the correct module. I think we want to enforce a basic structure as to how genesis information is passed from the genesis block directly to the module to handle. Right now we simply pass a json.rawmessage contained by a key with the module with the matching name, the module can parse / handle the raw message in arbitrary ways. I don't really see the advantage of customizable global application init genesis method... right now the only reason this is necessary is you are extending the standard account. Anyways let's talk more about this

Reply:
> What makes InitGenesis different from BeginBlock and EndBlock that it gets this special treatment ?

we need to route genesis information based on the key provided in genesis. BeginBlock/EndBlock do not need to route custom information - I also assume that BeginBlock/EndBlock handlers may have some interdependence whereas InitGenesis will not

further thoughts? - maybe we should break this out into a discussion issue

Reply:
@rigelrozanski Discussion issue: https://github.com/cosmos/cosmos-sdk/issues/759